solidity/contracts/bridge/Wallets.sol

```diff
@@ -305,6 +327,145 @@ library Wallets {
+        Data storage self,
+        bytes20 walletPubKeyHash
+    ) external {
+        // TODO: Perform slashing of wallet operators.
+
+        WalletState walletState = self
+            .registeredWallets[walletPubKeyHash]
+            .state;
+
+        require(walletState != WalletState.Unknown, "Unknown wallet");
+
+        if (walletState == WalletState.Live) {
+            requestWalletClosure(self, walletPubKeyHash);
+        }
+    }
+
+    /// @notice Handles a notification about wallet exhaustion and triggers the
+    ///         wallet closure process.
+    /// @param walletPubKeyHash 20-byte public key hash of the wallet
+    /// @param walletMainUtxo Data of the wallet's main UTXO, as currently
+    ///        known on the Ethereum chain.
+    /// @dev Requirements:
+    ///      - Wallet must not be set as the current active wallet
+    ///      - Wallet must exceed the wallet maximum age OR the wallet BTC
+    ///        balance must be lesser than the minimum threshold. If the latter
+    ///        case is true, the `walletMainUtxo` components must point to the
+    ///        recent main UTXO of the given wallet, as currently known on the
+    ///        Ethereum chain. If the wallet has no main UTXO, this parameter
+    ///        can be empty as it is ignored since the wallet balance is
+    ///        assumed to be zero.
+    ///      - Wallet must be in Live state
+    function notifyWalletExhausted(
+        Data storage self,
+        bytes20 walletPubKeyHash,
+        BitcoinTx.UTXO calldata walletMainUtxo
+    ) external {
+        require(
+            self.activeWalletPubKeyHash != walletPubKeyHash,
+            "Active wallet cannot be considered exhausted"
+        );
+
+        /* solhint-disable-next-line not-rely-on-time */
+        bool walletMaxAge = block.timestamp >=
+            self.registeredWallets[walletPubKeyHash].createdAt + self.maxAge;
+
+        require(
+            walletMaxAge ||
+                getWalletBtcBalance(self, walletPubKeyHash, walletMainUtxo) <
+                self.minBtcBalance,
+            "Wallet exhaustion conditions are not met"
+        );
+
+        requestWalletClosure(self, walletPubKeyHash);
+    }
+
+    /// @notice Requests wallet closure. If the wallet balance is greater than
+    ///         zero, the wallet is requested to move their funds first.
+    ///         Otherwise, the wallet is closed immediately and the ECDSA
+    ///         registry is notified about this fact. If the wallet closure
+    ///         request refers to the current active wallet, such a wallet
+    ///         is no longer considered active and the active wallet slot
+    ///         is unset allowing to trigger a new wallet creation immediately.
+    /// @param walletPubKeyHash 20-byte public key hash of the wallet
+    /// @dev Requirements:
+    ///      - Wallet must be in Live state
+    function requestWalletClosure(Data storage self, bytes20 walletPubKeyHash)
```

Comment:
I think we should rename this function to `moveFunds`. It should simplify state transition: wallet is `Live` and needs to get Closed. OK, the first step is always to `MoveFunds`. If the wallet has no funds to move, we can use a shortcut and mark it as `Closed`. Otherwise, we always wait for the confirmation.

```
[Live] -- moveFunds() --> [MovingFunds] -- notifyFundsMoved() --> [Closed]
[Live] -- moveFunds(); it is already empty!  --> [Closed]
```

Reply:
See 52889324